[UnifiedPDF] Support radio buttons, checkboxes, and choice annotations.

sammygill opened the pull request:
#### dceee48d7d3f9528cc05346a2a765fd7588485bc
<pre>
[UnifiedPDF] Support radio buttons, checkboxes, and choice annotations.
<a href="https://bugs.webkit.org/show_bug.cgi?id=268196">https://bugs.webkit.org/show_bug.cgi?id=268196</a>
<a href="https://rdar.apple.com/problem/121692745">rdar://problem/121692745</a>

Reviewed by Simon Fraser.

Choice annotations are similar to text widget annotations, which we
already support, in that they use a PDFPluginAnnotation to support
user interaction and just commit the result to the underlying annotation
afterwards.

Radio buttons and checkboxes involve slightly more work as we must keep
track of the mouse as the user interacts with them. Holding the left
mouse button on them will create a highlight effect indicating that this
action is occuring and releasing it will trigger the effect of the
checkbox/radio button. If the mouse is dragged off while the left mouse
button is held down on it, then the highlight disappears and no effect
should occur when the button is released. In order to accomplish this,
we need to start tracking the annotation that is being interacted with
through various mouse events. m_trackedAnnotation was added for this
purpose and is updated through the various events.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::annotationForRootViewPoint const):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::startAnnotationTracking):
(WebKit::UnifiedPDFPlugin::finishAnnotationTracking):
(WebKit::UnifiedPDFPlugin::handleMouseDraggedOffTrackedAnnotation):

Canonical link: <a href="https://commits.webkit.org/273921@main">https://commits.webkit.org/273921@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7fccc1a2f33b3895ce92409c6924b631beeb44bd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37007 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15919 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39322 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39599 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33084 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18458 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13049 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31605 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37568 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13425 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32640 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11709 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11724 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40852 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33529 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33280 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37624 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12028 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9831 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35754 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13677 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8405 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12406 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12941 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
IMO, let's be consistent across the tracked annotation and the active annotation? Give them both the PDFPluginAnnotation suffix or neither?

```
m_trackedAnnotation was added for this
purpose and is updated through the various events. m_activeAnnotation
was renamed to m_activePDFPluginAnnotation to distinguish it by strongly
indicating the thing that is active is a PDFPluginAnnotation.
```

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -806,6 +806,17 @@
+std::optional<RetainPtr<PDFAnnotation>> UnifiedPDFPlugin::annotationForRootViewPoint(const WebCore::IntPoint& point) const
+{
+    auto pointInDocumentSpace = convertFromPluginToDocument(convertFromRootViewToPlugin(point));
+    auto nearestPageIndex = nearestPageIndexForDocumentPoint(pointInDocumentSpace);
```
I think I've seen this pattern enough times to warrant us eventually making something like this:

```cpp
class UnifiedPDFPlugin:
enum class CoordinateSpace : uint8_t { RootView, Plugin, Contents, Document, Page };
std::optional<PDFDocumentLayout::PageIndex> nearestPageIndexForPoint(const WebCore::IntPoint&, CoordinateSpace);

auto UnifiedPDFPlugin::annotationForRootViewPoint(auto& point) const
{
    auto nearestPageIndex = nearestPageIndexForPoint(point, CoordinateSpace::RootView);
```

aprotyas replied:
(Out of scope for this PR, but I wanted to write it down for my own reference!)

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -806,6 +806,17 @@
+        return std::nullopt;
+
+    auto page = m_documentLayout.pageAtIndex(nearestPageIndex.value());
+    return [page annotationAtPoint: convertFromDocumentToPage(pointInDocumentSpace, nearestPageIndex.value())];
```
No space in objC calls

```suggestion
    return [page annotationAtPoint:convertFromDocumentToPage(pointInDocumentSpace, nearestPageIndex.value())];
```

sammygill replied:
Fixed! I wish the styler checker would catch these :(

aprotyas replied:
+1, could you file a bugzilla?

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -214,6 +217,7 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
 
     WebCore::IntPoint convertFromPluginToDocument(const WebCore::IntPoint&) const;
     std::optional<PDFDocumentLayout::PageIndex> nearestPageIndexForDocumentPoint(const WebCore::IntPoint&) const;
+    std::optional<RetainPtr<PDFAnnotation>> annotationForRootViewPoint(const WebCore::IntPoint&) const;
```
A pointer type has nullness built in, do we need to wrap it in `std::optional`?

aprotyas replied:
or do you intend for there to be a semantic difference between the `std::nullopt` and `nullptr` return values here?

sammygill replied:
Yeah I was probably overcomplicating it. I wanted to really make sure the caller checks that the `RetainPtr` could very easily be null in this case, but the function definition makes it clear enough

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -909,28 +920,49 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
-            auto pointInDocumentSpace = convertFromPluginToDocument(convertFromRootViewToPlugin(event.position()));
-            auto nearestPageIndex = nearestPageIndexForDocumentPoint(pointInDocumentSpace);
-            if (!nearestPageIndex)
+            if (RetainPtr<PDFAnnotation> annotation = annotationForRootViewPoint(event.position()).value_or(nullptr)) {
```
If we didn't return `std::optional<...>` here we could just do this here
```suggestion
            if (RetainPtr<PDFAnnotation> annotation = annotationForRootViewPoint(event.position())) {
```

sammygill replied:
Done!

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -909,28 +920,49 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+                }
+                if ([annotation isKindOfClass:getPDFAnnotationButtonWidgetClass()]) {
+                    startAnnotationTracking(annotation);
+                    return true;
```
Pretty sure you'll need to suppress deprecation warnings for these.

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1168,32 +1200,72 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+            [m_trackedAnnotation setButtonWidgetState: PDFWidgetCellState::kPDFWidgetOffState];
+        else if (currentButtonState == PDFWidgetCellState::kPDFWidgetOffState)
+            [m_trackedAnnotation setButtonWidgetState: PDFWidgetCellState::kPDFWidgetOnState];
+    } else
+        ASSERT_NOT_IMPLEMENTED_YET();
```
Won't this fire too much 😛

sammygill replied:
It might! But perhaps it's ok in debug builds since this is an active construction site! Ok with removing if we think that's the better option instead

aprotyas replied:
🤷 eh, we'll see if it fires too much

aprotyas reviewed (commented):
Looks good mostly! There's some inconsistency in the spacing when you make ObjC function calls (there should be no space between argument names and values).

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.mm:
```diff
@@ -117,7 +117,7 @@
 {
     if (event.type() == eventNames().blurEvent || event.type() == eventNames().changeEvent) {
-        m_plugin->setActiveAnnotation(0);
+        m_plugin->setActivePDFPluginAnnotation(0);
```
Pass `nil` not `0`

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -806,6 +806,17 @@
+RetainPtr<PDFAnnotation> UnifiedPDFPlugin::annotationForRootViewPoint(const WebCore::IntPoint& point) const
+{
+    auto pointInDocumentSpace = convertFromPluginToDocument(convertFromRootViewToPlugin(point));
+    auto nearestPageIndex = nearestPageIndexForDocumentPoint(pointInDocumentSpace);
```
This will conflict soon.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -909,28 +920,49 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
-            auto pointInDocumentSpace = convertFromPluginToDocument(convertFromRootViewToPlugin(event.position()));
-            auto nearestPageIndex = nearestPageIndexForDocumentPoint(pointInDocumentSpace);
-            if (!nearestPageIndex)
+            if (RetainPtr<PDFAnnotation> annotation = annotationForRootViewPoint(event.position())) {
```
Does the coordinate conversion work for iframe PDFs?

sammygill replied:
It looks like it does for the most part, but it seems to breakdown a bit when decreasing the size of iframes. Since this was a refactoring of some already existing logic into a function I'll create a follow up patch for it

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -909,28 +920,49 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
         default:
             return false;
         }
+    case WebEventType::MouseUp:
```
What if you click in an iframe PDF and then drag out. Do we miss the MouseUp? If we do, does it leave us in a bad state?

sammygill replied:
It does seem to leave us in a bad state (even happens with main frame PDFs on shipping)! To help get us out of this bad state, I added some logic that will check if there is a tracked annotation on a mouse move and treat it as if the cursor was dragged off the annotation

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1168,32 +1200,72 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
 #endif
 }
 
+void UnifiedPDFPlugin::startAnnotationTracking(RetainPtr<PDFAnnotation>& annotation)
```
It would be clearer if the argument were a `RetainPtr<PDFAnnotation>&&` that it's not an out param.

sammygill replied:
Fixed

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
```diff
@@ -86,7 +86,7 @@ class PDFPlugin final : public PDFPluginBase {
 
     void paintControlForLayerInContext(CALayer *, CGContextRef);
-    void setActiveAnnotation(RetainPtr<PDFAnnotation>&&) final;
+    void setActivePDFPluginAnnotation(RetainPtr<PDFAnnotation>&&) final;
```
I'm not convinced by the naming; also `PDFPluginAnnotation` implies that it belongs to the `PDFPlugin` class which isn't true in the UnifiedPDF code.

sammygill replied:
I'll modify the naming so that it better represents that this is a HTML annotation. I'll leave any renaming out of this patch and have a dedicated renaming patch since I'll want to modify the class names also

sammygill committed:
[UnifiedPDF] Support radio buttons, checkboxes, and choice annotations.
https://bugs.webkit.org/show_bug.cgi?id=268196
rdar://problem/121692745

Reviewed by Simon Fraser.

Choice annotations are similar to text widget annotations, which we
already support, in that they use a PDFPluginAnnotation to support
user interaction and just commit the result to the underlying annotation
afterwards.

Radio buttons and checkboxes involve slightly more work as we must keep
track of the mouse as the user interacts with them. Holding the left
mouse button on them will create a highlight effect indicating that this
action is occuring and releasing it will trigger the effect of the
checkbox/radio button. If the mouse is dragged off while the left mouse
button is held down on it, then the highlight disappears and no effect
should occur when the button is released. In order to accomplish this,
we need to start tracking the annotation that is being interacted with
through various mouse events. m_trackedAnnotation was added for this
purpose and is updated through the various events.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::annotationForRootViewPoint const):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::startAnnotationTracking):
(WebKit::UnifiedPDFPlugin::finishAnnotationTracking):
(WebKit::UnifiedPDFPlugin::handleMouseDraggedOffTrackedAnnotation):

Canonical link: https://commits.webkit.org/273921@main
